c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+                    continue;
+                }
+                &parent.children
+            } else {
+                &out
+            };
+
+            if let Some(prev) = siblings.last() {
+                debug_assert!(prev.span.lo() <= span.lo());
+                if prev.span.overlaps(span) {
+                    errs.push((span, rw, RewriteError::PartialOverlap(prev.span)));
+                    continue;
+                }
+            }
+
+            // TODO: check that all children are emitted somewhere in the `Rewrite`
+            // (e.g. given the rewrite `f(x + y) -> x`, we can't also have a rewrite on `x + y`,
+            // because that expression is discarded as part of the outer rewrite)
+            //
+            // Specifically: if there is a parent, check that either the current item's span is
+            // contained in the parent span and the parent's rewrite contains `Rewrite::Identity`,
+            // or the current item's span is contained in the span of some `Rewrite::Sub` in
+            // the parent rewrite.  If this doesn't hold, then produce `RewriteError::Discarded`.
+
+            // Push a new node onto the stack.
+            stack.push(RewriteTree {
+                span,
+                rw,
+                children: Vec::new(),
+            });
+        }
+
+        while let Some(rt) = stack.pop() {
+            commit(&mut stack, &mut out, rt);
+        }
+
+        (out, errs)
+    }
+}
+
+/// Split `rts` into the portion before `span`, the portion overlapping `span`, and the portion
+/// after `span`.  Nodes within `rts` should not overlap each other, and the list must be sorted by
+/// span; otherwise, the result are unspecified.
+fn partition_nodes<'a>(
+    rts: &'a [RewriteTree],
+    span: Span,
+) -> (&'a [RewriteTree], &'a [RewriteTree], &'a [RewriteTree]) {
+    let lo = span.lo();
+    let hi = span.hi();
+
+    // Collect nodes from the front of `rts` until we find one that overlaps or comes after `span`.
+    let i = rts
+        .iter()
+        .position(|rt| rt.span.hi() > lo)
+        .unwrap_or(rts.len());
+    let (before, rest) = rts.split_at(i);
+
+    // Collect nodes from the front of `rts` until we find one that comes strictly after `span`.
+    let j = rest
+        .iter()
+        .position(|rt| rt.span.lo() >= hi)
+        .unwrap_or(rest.len());
+    let (overlap, after) = rest.split_at(j);
+
+    (before, overlap, after)
+}
+
+struct Emitter<'a, F> {
+    file: &'a SourceFile,
+    emit: &'a mut F,
+}
+
+impl<'a, F: FnMut(&str)> Emitter<'a, F> {
+    fn emit_str(&mut self, s: &str) {
+        (self.emit)(s);
+    }
+
+    fn emit_parenthesized(&mut self, cond: bool, f: impl FnOnce(&mut Self)) {
+        if cond {
+            self.emit_str("(");
+        }
+        f(self);
+        if cond {
+            self.emit_str(")");
+        }
+    }
+
+    fn emit_rewrite(
+        &mut self,
+        rw: &Rewrite,
+        prec: usize,
```

Comment:
could you document `prec`?

Reply:
Why do we need precedence?  Can't we just parenthesize everything and let another tool remove the unnecessary ones?  We already emit a bunch of unnecessary operations.

Reply:
>Can't we just parenthesize everything and let another tool remove the unnecessary ones?

I tried implementing this, but it makes the output much harder to read.  Here's an example:
```diff
--- a/c2rust-analyze/tests/filecheck/insertion_sort_rewrites.rs
+++ b/c2rust-analyze/tests/filecheck/insertion_sort_rewrites.rs
@@ -9,20 +9,20 @@ extern crate libc;
 pub unsafe extern "C" fn insertion_sort(n: libc::c_int, p: *mut libc::c_int) {
     let mut i: libc::c_int = 1 as libc::c_int;
     while i < n {
-        // CHECK-DAG: [[#@LINE+2]]: p: &*$e
-        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): &(&$0[($1 as usize) ..])[0]
+        // CHECK-DAG: [[#@LINE+2]]: p: (&(*($e)))
+        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): (&((&(($0)[(($1) as usize) ..]))[(0)]))
         let tmp: libc::c_int = *p.offset(i as isize);
         let mut j: libc::c_int = i;
-        // CHECK-DAG: [[#@LINE+2]]: p: &*$e
-        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): &(&$0[($1 as usize) ..])[0]
+        // CHECK-DAG: [[#@LINE+2]]: p: (&(*($e)))
+        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): (&((&(($0)[(($1) as usize) ..]))[(0)]))
         while j > 0 as libc::c_int && *p.offset((j - 1 as libc::c_int) as isize) > tmp {
-            // CHECK-DAG: [[#@LINE+3]]: p: &*$e
-            // CHECK-DAG: [[#@LINE+2]]: p.offset({{.*}}): &(&$0[($1 as usize) ..])[0]
-            // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): &mut (&mut $0[($1 as usize) ..])[0]
+            // CHECK-DAG: [[#@LINE+3]]: p: (&(*($e)))
+            // CHECK-DAG: [[#@LINE+2]]: p.offset({{.*}}): (&((&(($0)[(($1) as usize) ..]))[(0)]))
+            // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): (&mut ((&mut (($0)[(($1) as usize) ..]))[(0)]))
             *p.offset(j as isize) = *p.offset((j - 1 as libc::c_int) as isize);
             j -= 1
         }
-        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): &mut (&mut $0[($1 as usize) ..])[0]
+        // CHECK-DAG: [[#@LINE+1]]: p.offset({{.*}}): (&mut ((&mut (($0)[(($1) as usize) ..]))[(0)]))
         *p.offset(j as isize) = tmp;
         i += 1
     }
```
I find it a lot easier to tell what's going on in the old versions and to check if I'm getting the outputs I was expecting.

I added some comments to explain the purpose of `prec` instead.

Reply:
That works I think.  I forgot we have to still interact with the rewrites themselves during testing, not just the final rewritten code after we can run `clippy` or `cargo check --fix` on it.